### ISSUE TITLE
NULL in Mutate as result of expression

### ISSUE THREAD

**Issue by aaronjg** [tidyverse/dplyr#2945]
If a column is set to NULL in mutate it is deleted. However, this does not work if mutate is returned as a result of an expression. It would also be nice if you could use NULL in summarize to exclude a column as well, but maybe that should go into a new feature request?

```r
> df <- data.frame(a=5,b=6) 
> df %>% mutate(a= NULL)
  b
1 6
> z <- NULL
> df %>% mutate(a=z)
Error: Unsupported type NILSXP for column "a
```

Previous discussion of the NULL feature. 
https://github.com/tidyverse/dplyr/issues/462
https://github.com/tidyverse/dplyr/commit/8104dd67705626d96bfcf299692ea7954cec8d3d

**Comment by krlmlr** [comment]
Thanks. This is by design, in particular because for a grouped `mutate()` it would be unclear what to do if the mutator returns `NULL` for some groups only. For `summarize()` we don't seem to need this, because the result includes only the grouping variables and those that are explicitly created.

**Comment by hadley** [comment]
Don't we have another open issue about this?

**Comment by krlmlr** [comment]
We decided to improve error messages only, but apparently we did this only for the ungrouped case...

**Comment by krlmlr** [comment]
I'm now getting:

``` r
library(dplyr)
z <- NULL
data.frame(a=5,b=6) %>% group_by(b) %>% mutate(a=z)
#> Error in mutate_impl(.data, dots): Column `a` is of unsupported type NULL
```

Same for ungrouped data.

@hadley: Do we need to add a special-case for the wording of the error message here?

**Comment by hadley** [comment]
The error message is ok for now. I'd rather make it possible to remove the column by setting it to `NULL`.

**Comment by krlmlr** [comment]
We don't remove because it's unclear what to do in a grouped call when it's removed only for some groups:

```r
tibble(a = 1:3, b = 4:6) %>% group_by(a) %>% mutate(b = if (a < 2) NULL else b)
```

**Comment by hadley** [comment]
Right, but that's mostly an artificial example, and could presumably be handled in the same way as 

```R
tibble(a = 1:3, b = 4:6) %>%
  group_by(a) %>%
  mutate(b = if (a < 2) 1 else "b")
```

**Comment by aaronjg** [comment]
But similar issues can happen with differing group lengths, and in that case, an error is thrown only when the lengths are invalid:

   > tibble(a = 1:3, b = 4:6) %>% group_by(a) %>% mutate(b = if (a < 2) rep(1,b) else b)
   Error in mutate_impl(.data, dots) : 
      Column `b` must be length 1 (the group size), not 4

**Comment by krlmlr** [comment]
We should postpone this until after #2311.

**Comment by hadley** [comment]
Agreed.
